configdialog.cpp

```diff
@@ -518,6 +535,78 @@ void PaletteCustomizeDialog::get_and_save_color(int palette_index) {
 	}
 }
 
+std::wstring PaletteCustomizeDialog::export_palettes() {
+	auto all_palettes = m_palette_repo.get_all_custom_palettes();
+	std::wstring palettes_string = L"";
+
+	for (const auto &palette : all_palettes) {
+		auto name = palette.name;
+		auto *data = palette.data;
+
+		palettes_string.append(palette.name + L": ");
+		palettes_string.append(m_palette_repo.serialize(*data));
+		palettes_string.append(L";\r\n");
+	}
+	palettes_string.append(L"~");
+
+	return palettes_string;
+}
+
+void PaletteCustomizeDialog::import_palettes(std::wstring *palettes) {
+	static std::wstring valid_chars = L"qwertyuiopasdfghjklzxcvbnmQWERTYUIOPASDFGHJKLZXCVBNM0123456789-_!?:";
+	static std::wstring valid_hex_chars = L"0123456789abcdefABCDEF";
+	size_t current_index = palettes->find_first_of(valid_chars.c_str());
+	std::wstring last_successful_palette_name;
+	std::wstring current_name;
+	std::wstring current_data[7];
+	bool graceful_exit = false;
+	try {
+		while (true) {
+			current_name = palettes->substr(current_index, palettes->find_first_of(L':', current_index) - current_index);
+			if (current_name.find_first_not_of(valid_chars + L" ") != std::wstring::npos) {
+				throw ERROR_BAD_ARGUMENTS;
+			}
+			for (int i = 0; i < 7; i++) {
+				current_index = palettes->find_first_of(L'#', current_index) + 1;
+				if (palettes->find_first_of(L';', current_index) - current_index != 6) {
+					throw ERROR_BAD_ARGUMENTS;
+				}
+				current_data[i] = palettes->substr(current_index, 6);
+				if (current_data[i].find_first_not_of(valid_hex_chars) != std::wstring::npos) {
+					throw ERROR_BAD_ARGUMENTS;
+				}
+				current_index = palettes->find_first_of(L';', current_index);
+			}
+			last_successful_palette_name = current_name;
+			if (palettes->find_first_of(valid_chars, current_index) == std::wstring::npos) {
+				if (palettes->find_first_of(L'~', current_index) != std::wstring::npos) {
+					graceful_exit = true;
+				}
+				break;
+			}
+			else {
+				current_index = palettes->find_first_of(valid_chars, current_index);
+			}
+		}
+		if (!graceful_exit) {
+			MessageBox(
+				NULL,
+				(L"A terminating character was not found. One or more palettes may have failed to import. Last successful import: " + last_successful_palette_name).c_str(),
+				L"Palette Customizer",
+				MB_ICONWARNING
+			);
+		}
+	}
+	catch (...) {
+		MessageBox(
+			NULL,
+			(L"An error occurred when importing palette \"" + current_name + L"\". Last successful import: \"" + last_successful_palette_name + L"\"").c_str(),
```

Comment:
To expand on the last comment, I think it should be possible to avoid showing users an error message altogether. Error boxes are annoying - they make a stupid noise, rip control away from your program, say some dumb technical jargon that means nothing to you, and then leave you confused what to do with the information it tried to tell you but you probably didn't read because it's been a long day and you're tired and you just want the fucking god forsaken thing to do the fucking thing it's fucking supposed to.

In my opinion, if the import is going to fail, then you just shouldn't be able to push the button at all. The Import button should only be active if the string you pasted in is OK. So, this function could be adjusted to just return a "bad" signal every time it would throw, and if everything goes alright then it returns the palettes that it parsed. Then rather than being called whenever someone pushes "Import," it's called every time the input box changes, with the results being stored somewhere if they're OK and used once "Import" is actually pressed.

(This is _exactly_ the use case for `std::optional`. I think a perfect return type for this function might be `std::optional<std::vector<PaletteData>>`.)